### PR TITLE
commit all

### PR DESCRIPTION
**Pull request by yyyogev**

### CODE REVIEW

File: src/main/java/io/logz/jmx2logzio/clients/JolokiaClient.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+            return metrics;
+        } catch (IOException e) {
+            throw new MBeanClientPollingFailure("Failed reading beans from Jolokia. Error = "+e.getMessage(), e);
+        }
+    }
+
+    private Dimension stringArgToDimension(String arg){
+        String[] argKeyVale = arg.split("=");
+        Dimension dimension = new Dimension(argKeyVale[0], argKeyVale.length > 1 ? argKeyVale[1] : "");
```
Comment (idohalevi) [inline src/main/java/io/logz/jmx2logzio/clients/JolokiaClient.java]: give meaningful names to the indexes

File: src/main/java/io/logz/jmx2logzio/clients/JolokiaClient.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+            if (value instanceof Map) {
+                Map<String, Number> flattenValueTree = flatten((Map) value);
+
+                for (String internalMetricName : flattenValueTree.keySet()) {
+                    metricValues.put(
+                            sanitizeMetricName(key, /*keepDot*/ false) + "."
+                                    + sanitizeMetricName(internalMetricName, /*keepDot*/ false),
+                            flattenValueTree.get(internalMetricName));
+                }
+            } else {
```
Comment (idohalevi) [inline src/main/java/io/logz/jmx2logzio/clients/JolokiaClient.java]: else if

File: src/main/java/io/logz/jmx2logzio/clients/ListenerWriter.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+public class ListenerWriter implements Shutdownable {
+    private static final Logger logger = LoggerFactory.getLogger(ListenerWriter.class);
+
+    private final static ObjectMapper mapper = new ObjectMapper();
+    private final BlockingQueue<Metric> messageQueue;
+    private ScheduledExecutorService scheduledExecutorService;
+    private HttpsRequestConfiguration requestConf;
+    private LogzioJavaSenderParams logzioSenderParams;
+    private LogzioSender logzioSender;
```
Comment (idohalevi) [inline src/main/java/io/logz/jmx2logzio/clients/ListenerWriter.java]: finals??

File: src/main/java/io/logz/jmx2logzio/clients/ListenerWriter.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+    public ListenerWriter(Jmx2LogzioConfiguration requestConf) {
+
+        this.logzioSenderParams = requestConf.getSenderParams();
+        initLogzioSender();
+        messageQueue = new LinkedBlockingQueue<>();
+
+        scheduledExecutorService = Executors.newScheduledThreadPool(1,
```
Comment (idohalevi) [inline src/main/java/io/logz/jmx2logzio/clients/ListenerWriter.java]: new single thread like above

File: src/main/java/io/logz/jmx2logzio/clients/ListenerWriter.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        } else {
+            senderBuilder.withInMemoryQueue()
+                    .setCapacityInBytes(logzioSenderParams.getInMemoryQueueCapacityInBytes())
+                    .setLogsCountLimit(logzioSenderParams.getLogsCountLimit())
+                    .endInMemoryQueue();
+        }
+        try {
+            this.logzioSender = senderBuilder.build();
+        } catch (LogzioParameterErrorException e) {
+            logger.error("promblem in one or more parameters with error {}", e.getMessage());
```
Comment (idohalevi) [inline src/main/java/io/logz/jmx2logzio/clients/ListenerWriter.java]: typo `promblem` + throw exception

File: src/main/java/io/logz/jmx2logzio/clients/ListenerWriter.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        }
+        this.logzioSender.start();
+    }
+
+    private SenderStatusReporter getStatusReporter() {
```
Comment (idohalevi) [inline src/main/java/io/logz/jmx2logzio/clients/ListenerWriter.java]: take `SenderStatusReporter ` to a different class and use it as a member

File: src/main/java/io/logz/jmx2logzio/clients/ListenerWriter.java [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+                logger.error(s);
+            }
+
+            @Override
+            public void warning(String s) {
+                logger.warn(s);
+            }
+
+            @Override
+            public void warning(String s, Throwable throwable) {
```
Comment (idohalevi) [inline src/main/java/io/logz/jmx2logzio/clients/ListenerWriter.java]: also add throwable when given

File: src/main/java/io/logz/jmx2logzio/configuration/Jmx2LogzioConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+
+    // Which client should we use
+    private MetricClientType metricClientType;
+
+    public enum MetricClientType {
+        JOLOKIA,
+        MBEAN_PLATFORM
+    }
+
+    public Jmx2LogzioConfiguration(Config config) throws IllegalConfiguration {
```
Comment (idohalevi) [inline src/main/java/io/logz/jmx2logzio/configuration/Jmx2LogzioConfiguration.java]: `Jmx2LogzioConfiguration` too long and hard to follow. Take out to smaller functions..

File: src/main/java/io/logz/jmx2logzio/configuration/Jmx2LogzioConfiguration.java [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+            serviceHost = config.getString(Jmx2LogzioJavaAgent.SERVICE_HOST);
+        }
+
+        if (config.hasPath(JolokiaClient.POLLER_JOLOKIA)) {
+            metricClientType = MetricClientType.JOLOKIA;
+        } else if (config.hasPath(Jmx2LogzioJavaAgent.POLLER_MBEAN_DIRECT)) {
+            metricClientType = MetricClientType.MBEAN_PLATFORM;
+        }
+
+        if (this.metricClientType == MetricClientType.JOLOKIA) {
```
Comment (idohalevi) [inline src/main/java/io/logz/jmx2logzio/configuration/Jmx2LogzioConfiguration.java]: can be inside `if (config.hasPath(JolokiaClient.POLLER_JOLOKIA)`? also what if both ifs are false?

File: src/main/java/io/logz/jmx2logzio/configuration/Jmx2LogzioConfiguration.java [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+            throw new IllegalConfiguration("please enter a valid logz.io token (can be located at apps.logz.io >> settings >> general)");
+        }
+        logzioJavaSenderParams.setToken(config.getString(LogzioJavaSenderParams.LOGZIO_TOKEN));
+
+        logzioJavaSenderParams.setFromDisk(config.hasPath(LogzioJavaSenderParams.FROM_DISK) ?
+                config.getBoolean(LogzioJavaSenderParams.FROM_DISK) : logzioJavaSenderParams.isFromDisk());
+
+        if (config.hasPath(LogzioJavaSenderParams.IN_MEMORY_QUEUE_CAPACITY)) {
+            int capacity = config.getInt(LogzioJavaSenderParams.IN_MEMORY_QUEUE_CAPACITY);
+            if (capacity > 0) {
```
Comment (idohalevi) [inline src/main/java/io/logz/jmx2logzio/configuration/Jmx2LogzioConfiguration.java]: trinary condition

File: src/main/java/io/logz/jmx2logzio/configuration/Jmx2LogzioConfiguration.java [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+                logzioJavaSenderParams.setInMemoryQueueCapacityInBytes(capacity);
+            } else {
+                logger.error("argument IN_MEMORY_QUEUE_CAPACITY has to be a natural number, using default instead: {}",logzioJavaSenderParams.getInMemoryQueueCapacityInBytes());
+            }
+        }
+
+        if (config.hasPath(LogzioJavaSenderParams.LOGS_COUNT_LIMIT)) {
+            int logCountLimit = config.getInt(LogzioJavaSenderParams.LOGS_COUNT_LIMIT);
+            if (logCountLimit > 0) {
```
Comment (idohalevi) [inline src/main/java/io/logz/jmx2logzio/configuration/Jmx2LogzioConfiguration.java]: trinary condition

File: src/main/java/io/logz/jmx2logzio/configuration/Jmx2LogzioConfiguration.java [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+                logzioJavaSenderParams.setLogsCountLimit(logCountLimit);
+            } else {
+                logger.error("argument LOGS_COUNT_LIMIT has to be a natural number, using default instead: {}", logzioJavaSenderParams.getLogsCountLimit());
+            }
+        }
+
+        if (config.hasPath(LogzioJavaSenderParams.DISK_SPACE_CHECK_INTERVAL)) {
+            int interval = config.getInt(LogzioJavaSenderParams.DISK_SPACE_CHECK_INTERVAL);
+            if (interval > 0) {
+                logzioJavaSenderParams.setDiskSpaceCheckInterval(interval);
```
Comment (idohalevi) [inline src/main/java/io/logz/jmx2logzio/configuration/Jmx2LogzioConfiguration.java]: trinary condition

File: src/main/java/io/logz/jmx2logzio/configuration/Jmx2LogzioConfiguration.java [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+        }
+
+        if (config.hasPath(LogzioJavaSenderParams.QUEUE_DIR)) {
+            File queuePath = new File(config.getString(LogzioJavaSenderParams.QUEUE_DIR));
+            logzioJavaSenderParams.setQueueDir(queuePath);
+        }
+
+        if (config.hasPath(LogzioJavaSenderParams.FILE_SYSTEM_SPACE_LIMIT)) {
+            int spaceLimit = config.getInt(LogzioJavaSenderParams.FILE_SYSTEM_SPACE_LIMIT);
+            if (spaceLimit > 0) {
```
Comment (idohalevi) [inline src/main/java/io/logz/jmx2logzio/configuration/Jmx2LogzioConfiguration.java]: trinary condition

File: src/test/java/io/logz/jmx2logzio/LogzioSenderTest.java [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+        writer.writeMetrics(metrics);
+        waitToSend();
+        recordedRequests = mockServerClient.retrieveRecordedRequests(request().withMethod("POST"));
+        Assert.assertEquals(recordedRequests.length, 1);
+        String message = recordedRequests[0].getBodyAsString();
+        Assert.assertTrue(message.contains("\"" + key + "\":" + value));
+        writer.shutdown();
+    }
+
+    @AfterTest
```
Comment (idohalevi) [inline src/test/java/io/logz/jmx2logzio/LogzioSenderTest.java]: Put `@AfterTest` after `@BeforeTest` for easier reading.

File: src/test/java/io/logz/jmx2logzio/LogzioSenderTest.java [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+    public void stopMockServer() {
```
Comment (idohalevi) [inline src/test/java/io/logz/jmx2logzio/LogzioSenderTest.java]: also put at the beginning

File: src/test/java/io/logz/jmx2logzio/Utils/MetricsPipelineTest.java [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+        attr.add("First measure");
+        attr.add(("Second measure"));
+        beans.add(new MetricBean("minMemoryUsagePercent", attr));
+        beans.add(new MetricBean("MaxCPUUsage", attr));
+        beans.add(new MetricBean("MaxMemoryUsagePercent", attr));
+        return metricsPipeline.getFilteredBeans(beans);
+    }
+
+    @AfterTest
+    private void clean() {
```
Comment (idohalevi) [inline src/test/java/io/logz/jmx2logzio/Utils/MetricsPipelineTest.java]: same as in `LogzioSenderTest.java`

File: src/main/java/io/logz/jmx2logzio/Jmx2Logzio.java [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+    private static final Logger logger = LoggerFactory.getLogger(Jmx2Logzio.class);
+
+    private final Jmx2LogzioConfiguration conf;
+    private final ScheduledThreadPoolExecutor taskScheduler;
+    private final MBeanClient client;
+
+    public Jmx2Logzio(Jmx2LogzioConfiguration conf) {
+        this.conf = conf;
+
+        this.taskScheduler = new ScheduledThreadPoolExecutor(1);
```
[the remark this answers is not in the document]
Comment (yyyogev) [reply src/main/java/io/logz/jmx2logzio/Jmx2Logzio.java]: Cool! didn't know about that method

File: src/main/java/io/logz/jmx2logzio/Jmx2Logzio.java [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+
+        if (conf.getMetricClientType() == JOLOKIA) {
+            this.client = new JolokiaClient(conf.getJolokiaFullUrl());
+            logger.info("Running with Jolokia URL: {}", conf.getJolokiaFullUrl());
+
+        } else if (conf.getMetricClientType() == MBEAN_PLATFORM) {
+            this.client = new JavaAgentClient();
+            logger.info("Running with Mbean client");
+        } else {
+            throw new IllegalConfiguration("Unsupported client type: " + conf.getMetricClientType());
```
[the remark this answers is not in the document]
Comment (yyyogev) [reply src/main/java/io/logz/jmx2logzio/Jmx2Logzio.java]: Originally, if the user used Jolokia agent, then he should have set the poller property in the configuration file. Since I changed that and wrote all the properties in the config file, the only case which it can happen is if the user will comment out the poller property in the conf file (and then it will be null). 
Is that too dummy proof? should I use ternary instead?

File: src/main/java/io/logz/jmx2logzio/Jmx2LogzioJavaAgent.java [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+        // Merge the two configurations
+        return userConfig.withFallback(fileConfig);
+    }
+
+    private static Map<String, String> parseArgumentsString(String arguments) throws IllegalConfiguration {
+        try {
+            Map<String, String> argumentsMap = new HashMap<>();
+            Map<String, String> keyValues = Splitter.on(',').omitEmptyStrings().withKeyValueSeparator('=').split(arguments);
+
+            keyValues.forEach((k, v) ->
```
[the remark this answers is not in the document]
Comment (yyyogev) [reply src/main/java/io/logz/jmx2logzio/Jmx2LogzioJavaAgent.java]: You mean the lambda expression? And document where? in the readme?
I can also iterate over the entrySet instead, whatever you think is best

[the remark this answers is not in the document]
Comment (yyyogev) [reply src/main/java/io/logz/jmx2logzio/Jmx2LogzioJavaAgent.java]: Just noticed that there are other lambda expressions usage in the project, so perhaps just documenting it is a better option

File: src/main/java/io/logz/jmx2logzio/clients/JavaAgentClient.java [hunk 20]
```diff
@@ -0,0 +1,206 @@
...
+            List<MetricBean> metricBeans = Lists.newArrayList();
+            Set<ObjectInstance> instances = server.queryMBeans(null, null);
+
+            for (ObjectInstance instance : instances) {
+                MBeanInfo mBeanInfo = server.getMBeanInfo(instance.getObjectName());
+                List<String> attributes = Lists.newArrayList();
+
+                for (MBeanAttributeInfo attribute : mBeanInfo.getAttributes()) {
+                    attributes.add(attribute.getName());
+                }
```
[the remark this answers is not in the document]
Comment (yyyogev) [reply src/main/java/io/logz/jmx2logzio/clients/JavaAgentClient.java]: Array don't have a stream() built in option. Should we use Arrays.stream(mBeanInfo.getAttributes()).foreach(...) instead?

File: src/main/java/io/logz/jmx2logzio/Jmx2Logzio.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (idohalevi) [reply src/main/java/io/logz/jmx2logzio/Jmx2Logzio.java]: @yyyogev Can you verify it's not null when you validate all the parameters? If so use trinary

File: src/main/java/io/logz/jmx2logzio/Jmx2LogzioJavaAgent.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (idohalevi) [reply src/main/java/io/logz/jmx2logzio/Jmx2LogzioJavaAgent.java]: Yes just document it in the readme. Ask @imnotashrimp to add it

File: src/main/java/io/logz/jmx2logzio/clients/JavaAgentClient.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (idohalevi) [reply src/main/java/io/logz/jmx2logzio/clients/JavaAgentClient.java]: You're right, no need.

File: src/main/resources/application.conf [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+  log-count-limit = -1
+  disk-space-checks-interval = 1000 //in milliseconds
+
+  queue-dir = "metrics"
+  file-system-full-percent-threshold = 98
+  clean-sent-metrics-interval = 30
+
+}
+
+metricsPollingIntervalInSeconds = 30
```
Comment (imnotashrimp) [inline src/main/resources/application.conf]: This is camelCase, other settings are kebab-case.

File: src/main/resources/application.conf [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+  poller {
+    white-list-regex = ".*"
+    black-list-regex = "$a"
+    jolokia {
+//      jolokiaFullUrl = "http://127.0.0.1:8778/jolokia/"
+    }
+  }
+}
+
+logzioJavaSender {
```
Comment (imnotashrimp) [inline src/main/resources/application.conf]: This is camelCase, other settings are kebab-case

File: src/main/java/io/logz/jmx2logzio/Jmx2Logzio.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (yyyogev) [reply src/main/java/io/logz/jmx2logzio/Jmx2Logzio.java]: Will do

File: src/main/java/io/logz/jmx2logzio/Jmx2LogzioJavaAgent.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (yyyogev) [reply src/main/java/io/logz/jmx2logzio/Jmx2LogzioJavaAgent.java]: @imnotashrimp please add Java 8 to the requirements in the Readme file

File: src/main/java/io/logz/jmx2logzio/clients/JolokiaClient.java [hunk 23]
```diff
@@ -0,0 +1,199 @@
...
+                        logger.info("Can't sent Metric since it's invalid: "+e.getMessage());
+                    }
+                }
+            }
+
+            return metrics;
+        } catch (IOException e) {
+            throw new MBeanClientPollingFailure("Failed reading beans from Jolokia. Error = "+e.getMessage(), e);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (yyyogev) [reply src/main/java/io/logz/jmx2logzio/clients/JolokiaClient.java]: @idohalevi did you mean getMetrics?

File: src/main/java/io/logz/jmx2logzio/configuration/Jmx2LogzioConfiguration.java (hunk 9 above)
Comment (yyyogev) [reply src/main/java/io/logz/jmx2logzio/configuration/Jmx2LogzioConfiguration.java]: Yes, I took care of it as you suggested above

File: src/main/java/io/logz/jmx2logzio/configuration/Jmx2LogzioConfiguration.java (hunk 10 above)
Comment (yyyogev) [reply src/main/java/io/logz/jmx2logzio/configuration/Jmx2LogzioConfiguration.java]: What's the motivation behind this? this isn't the traditional form of:
`if (expression) then
    var = a
  else 
    var = b `
if we'll use ternary condition it will look something like this:
`var = expression ? a : var`
which will cause a redundant placement.

File: src/main/java/io/logz/jmx2logzio/clients/JavaAgentClient.java [hunk 24]
```diff
@@ -0,0 +1,206 @@
...
+                metricValues.put(sanitizeMetricName(key, /*keepDot*/ false), (Number) value);
+            } else if (value instanceof CompositeData) {
+                CompositeData data = (CompositeData) value;
+                Map<String, Object> valueMap = handleCompositeData(data);
+                metricValues.putAll(prependKey(key, flatten(valueMap)));
+            } else if (value instanceof TabularData) {
+                TabularData tabularData = (TabularData) value;
+                Map<String, Object> rowKeyToRowData = handleTabularData(tabularData);
+                metricValues.putAll(prependKey(key, flatten(rowKeyToRowData)));
+            } else if (!(value instanceof String) && !(value instanceof Boolean)) {
```
[the remark this answers is not in the document]
Comment (yyyogev) [reply src/main/java/io/logz/jmx2logzio/clients/JavaAgentClient.java]: Not sure what to do with it, looks even more confusing now

**Comment by yyyogev** [comment]
@idohalevi all changes have been made, please review again

File: src/main/java/io/logz/jmx2logzio/clients/JolokiaClient.java [hunk 25]
```diff
@@ -87,73 +116,83 @@ public JolokiaClient(String jolokiaFullURL) {
...
+                ));
+            } catch (IllegalArgumentException e) {
+                logger.error("Can't send Metric since it's invalid: " + e.getMessage());
             }
+        }
+        return metrics;
+    }
 
-            return metrics;
+    private HttpResponse getHttpResponse(String jolokiaFullURL, String requestBody) {
```
Comment (idohalevi) [inline src/main/java/io/logz/jmx2logzio/clients/JolokiaClient.java]: It's unclear from the name what the function does. Maybe more like `sendToJolokia`?
